canduit.js

```diff
@@ -68,49 +77,80 @@ Canduit.prototype.parseConfigFile = function parseConfigFile (cb) {
   });
 };
 
-Canduit.prototype.exec = function exec (route, params, cb) {
-  var logger = this.logger;
+Canduit.prototype.exec = function exec(route, params, cb) {
+  var self = this;
 
-  if (this.session) {
-    params.__conduit__ = this.session;
+  if (!self.token) {
+    self.execSession(route, params, cb);
+  } else {
+    self.execToken(route, params, cb);
   }
+}
 
-  var req = request.post(this.api + route, {
+Canduit.prototype.execSession = function execSession(route, params, cb) {
+  var self = this;
+
+  if (self.session) {
+    params.__conduit__ = self.session;
+  }
+  var reqOptions = {
+    url: self.api + route,
     json: true,
     form: {
       output: 'json',
       params: JSON.stringify(params)
     }
-  }, function (err, response, data) {
-    if (err) return cb(err, null);
-    if (response.statusCode >= 400) {
-      return cb(Canduit.serverError(response), null);
-    }
-    if (data.error_info) {
-      return cb(Canduit.conduitError(data), null);
-    }
+  };
+  self.makeReq('post', reqOptions, cb);
+};
 
-    logger.log('%s responded with %s',
-      req.href, JSON.stringify(data));
+Canduit.prototype.execToken = function execToken(route, params, cb) {
+  var self = this;
+  var qs = params || {};
+  qs['api.token'] = self.token;
 
-    cb(null, data.result);
-  });
+  var reqOptions = {
+    url: self.api + route,
+    qs: qs,
+    json: true
+  };
+  self.makeReq('get', reqOptions, cb);
+};
 
-  logger.log('POST to %s with %s',
-    this.api + route, req.body.toString());
+Canduit.prototype.makeReq = function getOptionsCallback(type, reqOptions, cb) {
+  var self = this;
+  var req = request[type](reqOptions,
+    function getOptionsCallback(err, response, data) {
+      if (err) {
+        return cb(err, null);
+      }
+      if (response.statusCode >= 400) {
+        return cb(Canduit.serverError(response), null);
+      }
+      if (data.error_info) {
+        return cb(Canduit.conduitError(data), null);
+      }
+
+      self.logger.info('response from phabricator', {
+        href: req.href,
+        data: data
+      });
+
+      cb(null, data.result);
+    });
+  self.logger.info('request made to phabricator', reqOptions);
 };
 
-Canduit.prototype.authenticate = function authenticate (cb) {
-  if (!this.cert) return cb(null, this);
+Canduit.prototype.authenticate = function authenticate(cb) {
+  var self = this;
```

Comment:
declaring var self at the top clarifies when self is actually being declared due to hoisting, and using self throughout the function clarifies that the 'this' and 'self' are referring to the same this. I can change it back. This was something suggested to me.